src/main.py

```diff
@@ -0,0 +1,119 @@
+        
+        # Report loop results
+        def loop_callback(loop):
+            tests = loop.failed_tests()
+            if (len(tests) == 0):
+                m = "Loop #{} finished with success".format(loop.index)
+                step.add_log(tv.LogSeverity.INFO, m)
+            else:
+                # Log failed tests
+                names = ", ".join([t.name for t in tests])
+                m = "Loop #{} failed {} tests: {}".format(loop.index, len(tests), names)
+                step.add_error(symptom="loop-failure", message=m)
+                
+                # Record failed addresses
+                for test in tests:
+                    for r in test.result:
+                        bad_addrs.add_measurement(value=r.addr)
+        observer.callbacks.loop_ready = loop_callback
+        
+        # Report individual tests to make long runs more responsive
+        def test_callback(test):
+            if test.passed():
+                step.add_log(tv.LogSeverity.INFO, "Test '{}' passed".format(test.name))
+            else:
+                step.add_log(tv.LogSeverity.ERROR, "Test '{}' failed".format(test.name))
+        observer.callbacks.test_ready = test_callback
+        
+        # Report diagnosis
+        def run_callback(failed_loop_count):
+            if failed_loop_count == 0:
+                step.add_diagnosis(tv.DiagnosisType.PASS, verdict="memtester-passed")
+            else:
+                step.add_diagnosis(tv.DiagnosisType.FAIL, verdict="memtester-failed")
+                raise tv.TestRunError(status=tv.TestStatus.COMPLETE, result=tv.TestResult.FAIL)
+        observer.callbacks.run_ready = run_callback
+        
+        # Report parsing errors (supposed to only happen with unknown memtester versions)
+        def parsing_error_callback(desc):
+            step.add_error(symptom="memtester-parsing-error", message=desc)
+            raise tv.TestRunError(status=tv.TestStatus.ERROR, result=tv.TestResult.NOT_APPLICABLE)
+        observer.callbacks.parsing_error = parsing_error_callback
+
+        # Log raw memtester output if necessary
+        if args.mt_log_filename:
+            estack.enter_context(log_file := open(args.mt_log_filename, "w"))
+            def line_ready_callback(line):
+                log_file.write(line)
+            observer.callbacks.line_ready = line_ready_callback
+        
+        # Run memtester (finally!)
+        try:
+            mt_cmd = sh.Command(args.mt_path)
+            mt_args = args.mt_args.split(" ")
+            with contextlib.redirect_stderr(None): # Magic to silence sh lib
+                # For _ok_code see https://linux.die.net/man/8/memtester
+                observer.run(mt_cmd(*mt_args,
+                    _err_to_out=True, _iter=True, _ok_code=(0, 2, 4, 2 | 4)))
+        except sh.CommandNotFound as e:
+            m = "Memtester not found: {}".format(e)
+            step.add_error(symptom="memtester-not-found", message=m)
```

Comment:
i mentioned before, but this place is a bit far from the other comment; i think `memtester-not-found` is not a `tv.step` problem, but rather something that should be detected before starting any `tv.run`s

Reply:
I am confused. How do I report an error without starting a `tv.run`? According to the spec:
>Errors can be child artifacts of either TestRunArtifact or TestStepArtifact messages.

Reply:
here's an example of error before run start (this case shows issue in DUT discovery): https://github.com/opencomputeproject/ocp-diag-core-python/blob/dev/examples/simple.py#L97

Reply:
Thanks for the link! 

https://github.com/opencomputeproject/ocp-diag-memtester/issues/5